pilot/pkg/xds/endpoint_builder.go

```diff
@@ -370,3 +394,138 @@ func buildEnvoyLbEndpoint(e *model.IstioEndpoint) *endpoint.LbEndpoint {
 
 	return ep
 }
+
+// TODO this logic is probably done elsewhere in XDS, possible code-reuse + perf improvements
+type mtlsChecker struct {
+	push            *model.PushContext
+	svcPort         int
+	destinationRule *networkingapi.DestinationRule
+
+	// cache of host identifiers that have mTLS disabled
+	mtlsDisabledHosts map[string]struct{}
+
+	// cache of labels/port that have mTLS disabled by peerAuthn
+	peerAuthDisabledMTLS map[string]bool
+	// cache of labels that have mTLS disabled by a subset traffic policy
+	subsetPolicyDisabledMTLS map[string]bool
+	// true if the default traffic policy disables mTLS
+	disaledByDestinationRule bool
+}
+
+func newMtlsChecker(push *model.PushContext, svcPort int, dr *config.Config) *mtlsChecker {
+	var drSpec *networkingapi.DestinationRule
+	if dr != nil {
+		drSpec = dr.Spec.(*networkingapi.DestinationRule)
+	}
+	return &mtlsChecker{
+		push:                     push,
+		svcPort:                  svcPort,
+		destinationRule:          drSpec,
+		mtlsDisabledHosts:        map[string]struct{}{},
+		peerAuthDisabledMTLS:     map[string]bool{},
+		subsetPolicyDisabledMTLS: map[string]bool{},
+		disaledByDestinationRule: mtlsDisabledByDefaultTrafficPolicy(dr, svcPort),
+	}
+}
+
+// mTLSDisabled returns true if the given lbEp has mTLS disabled due to any of:
+// - disabled tlsMode
+// - DestinationRule disabling mTLS on the entire host or the port
+// - PeerAuthentication disabling mTLS at any applicable level (mesh, ns, workload, port)
+func (c *mtlsChecker) isMtlsDisabled(lbEp *endpoint.LbEndpoint) bool {
+	if c == nil {
+		return false
+	}
+	_, ok := c.mtlsDisabledHosts[lbEpKey(lbEp)]
+	return ok
+}
+
+// computeForEndpoint checks destination rule, peer authentication and metadata to determine if mTLS was turned off.
+// This must be done during conversion from IstioEndpoint since we still have workload metadata.
+func (c *mtlsChecker) computeForEndpoint(ep *model.IstioEndpoint) {
+	tlsMode := envoytransportSocketMetadata(ep.EnvoyEndpoint, model.TLSModeLabelShortname)
+	if tlsMode != model.IstioMutualTLSModeLabel ||
```

Comment:
we discussed offline that this may need to be inverted for when DR explicitly enables mTLS. I think we may need that before we ship 1.9.x with this or it could be a regression. Do you want to merge this now or wait for that?

Reply:
Right, I'll add that in master then include that in this PR.